cas-30 Update Proposal Card

CJamesKeller opened the pull request:
Depends on https://github.com/DapperCollectives/CAST/pull/13

germanurrus commented:
I think this one will be blocked by this PR https://github.com/DapperCollectives/CAST/pull/13 I'll try to finish the updates as soon as possible

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCard.js:
```diff
@@ -0,0 +1,87 @@
+  const cardRef = useRef();
+  const [cardWidth, setCardWidth] = useState();
+  useEffect(() => {
+    // used for wrapping proposal body text
```
Very much open to a better solution for this, if possible.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCardBody.js:
```diff
@@ -0,0 +1,39 @@
+const ProposalCardBody = ({ name, body, cardWidth, inlineImage }) => {
+  const cardPaddingOnly = 48;
+  const aLittleExtra = 2;
+  const noImageDefault = cardPaddingOnly + aLittleExtra;
```
Can change the naming here if too informal.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCardFooter.js:
```diff
@@ -0,0 +1,146 @@
+  }, [computedStatus, getVotingResults, id, justMounted]);
+
+  // prevents memory leak in above useEffect for getVotingResults
+  useEffect(() => () => { setJustMounted(false) }, []);
```
Open to better approaches, but was unable to resolve the memory leak without this.

germanurrus replied:
maybe triggering only by checking votingResults has something?

dbslone replied:
There is no need to wrap the custom hook for fetching voting results with useEffect. If you look at how `useVotingResults` is used in [ProposalInformation.js](https://github.com/DapperCollectives/CAST/blob/main/frontend/packages/client/src/components/ProposalInformation.js#L166-L167) you can just the following here to achieve the desired result:

```
const { data: votingResults } = useVotingResults(id);
```

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCardFooter.js:
```diff
@@ -0,0 +1,146 @@
+  };
+
+  const textDescriptionMap = {
+    [FilterValues.active]: `${diffDuration} left`,
```
Technically differs from the design, but more consistent and readable, in my opinion.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCardFooter.js:
```diff
@@ -0,0 +1,146 @@
+  };
+
+  const statusLabelMap = {
+    [FilterValues.active]: <StatusLabel status="Active" voted={voted} rounder
```
Note: We don't have any set line-length wrapping rules, but I wrapped attributes that would go over 100 chars, for readability, in all files. Can change this if preferred.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCardFooter.js:
```diff
@@ -0,0 +1,146 @@
+    [FilterValues.cancelled]: <StatusLabel status="Cancelled" voted={voted} rounder />,
+  };
+
+  // gross section to handle styling
```
Open to better approaches. It is gross.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCardHeader.test.js:
```diff
@@ -2,7 +2,7 @@ import React from "react";
 import "@testing-library/jest-dom";
 import renderer from "react-test-renderer";
-import ProposalHeader from "./ProposalHeader";
+import ProposalCardHeader from "./ProposalCardHeader";
```
**Be advised**: Component has been updated, but test has _not_.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalList.sass:
```diff
@@ -4,12 +4,23 @@
+  line-height: 140%
+  display: -webkit-box
+  -webkit-box-orient: vertical
+  -webkit-line-clamp: 3
```
To wrap and clip text in the proposal body.

CJamesKeller commented:
- Implements new proposal card layout for all viewport sizes.
- Extracts proposal card and sub-components into separate files, leaving as much as possible the same as it was previously.
- Adds `date-fns` and `string-strip-html` deps to parse dates and html.
- Applies lint fix.
- Removes unnecessary components.

germanurrus commented on frontend/packages/client/src/components/ProposalsList/ProposalCardFooter.js:
```diff
@@ -0,0 +1,146 @@
+  };
+
+  // gross section to handle styling
+  let extraClasses1 = "";
+  let extraClasses2 = "";
+  let extraStyles1 = {};
+  let extraClasses3 = "";
+  let extraClasses4 = "";
+
```
there's a WrapperComponent that handles styles from mobile and desktop https://github.com/DapperCollectives/CAST/blob/c37ce0380c3b9e71a72f7c753aed1b8f29b3d0ef/frontend/packages/client/src/components/WrapperResponsive.js 
maybe can be used to simplify the if clause

dbslone replied:
Anytime you find yourself in a situation where you use generic variable names like these it's a good signal to think a little more about your naming convention. Naming things is one of the hardest parts of programming. 

In this situation you could use names like `Container`, `Icon`, `Status`, `Content`, etc. Try to use a name that describes a little about the element where the class or style is being applied.

germanurrus commented on frontend/packages/client/src/components/ProposalsList/ProposalCardFooter.js:
```diff
@@ -0,0 +1,146 @@
+    [FilterValues.pending]: <StatusLabel status="Pending" voted={voted} rounder
+      color="has-background-grey-light"
+    />,
+    [FilterValues.closed]: <StatusLabel status="Cancelled" voted={voted} rounder />,
```
```suggestion
    [FilterValues.closed]: <StatusLabel status="Closed" voted={voted} rounder />,
```

dbslone commented on frontend/packages/client/src/components/ProposalsList/ProposalCardBody.js:
```diff
@@ -0,0 +1,39 @@
+  const imageSpacing = 10;
+  const inlineImageDefault = noImageDefault + imageBoxWidth + imageSpacing;
+
+  let maxWidth = 0;
+  if (!cardWidth) {
+    // default so initial result is 0 until actual width is known
+    maxWidth = inlineImage ? inlineImageDefault : noImageDefault;
+  } else {
+    maxWidth = cardWidth;
+  }
+  if (inlineImage) {
+    maxWidth -= inlineImageDefault;
+  } else {
+    maxWidth -= noImageDefault;
+  }
```
Is this needed? When I open the inspector and remove the maxWidth style from the body and adjust the window size I didn't notice any issues.

CJamesKeller commented:
@germanurrus @dbslone Removed width check in `ProposalCardBody`, removed redundant `useEffect` hook, corrected `statusLabelMap`, and implemented `<WrapperResponsive />`.

germanurrus commented on frontend/packages/client/src/components/ProposalsList/ProposalCardHeader.js:
```diff
@@ -0,0 +1,25 @@
+
+const ProposalCardHeader = ({ creatorAddr }) => {
+  const { communityId } = useParams();
+  const isAdmin = useUserRoleOnCommunity({
+    creatorAddr,
+    communityId,
+    roles: ["admin"],
+  });
```
I think this might not be efficient, and we can work it on a separate PR and [ticket](https://linear.app/dappercollectives/issue/CAS-88/improve-use-of-useuserroleoncommunity-for-proposal-list), but the userRoleOnCommunity is using the endpoint `/users/${addr}/communities?count=${count}&start=${start}` which will bring all the communities that the address is part of, then we are making the same request for all headers. I think we can move this out of the proposal card header, and pass it down as a prop. The request would be done inside `useCommunityProposalWithVotes` [here](https://github.com/DapperCollectives/CAST/blob/fb64ffde79e37522cf555c3deb98a2831988126c/frontend/packages/client/src/hooks/useCommunityProposalsWithVotes.js#L7-L8) this endpoint returns up to 100 elements, then at the top level we can filter which ids we have and set the prop as admin or not.

CJamesKeller replied:
Made a change for this in `<CommunityProposalList />`, and realized I had been using the wrong value. Let me know what you think.

germanurrus commented on frontend/packages/client/src/components/ProposalsList/ProposalCardHeader.js:
```diff
@@ -0,0 +1,25 @@
+
+const ProposalCardHeader = ({ creatorAddr }) => {
+  const { communityId } = useParams();
+  const isAdmin = useUserRoleOnCommunity({
+    creatorAddr,
+    communityId,
+    roles: ["admin"],
+  });
```
I think this might not be efficient, and we can work it on a separate PR, but the userRoleOnCommunity is using the endpoint `/users/${addr}/communities?count=${count}&start=${start}` which will bring all the communities that the address is part of, then we are making the same request for all headers. I think we can move this our of the proposal card header, and pass it down as a prop. The request would be done inside `useCommunityProposalWithVotes` [here](https://github.com/DapperCollectives/CAST/blob/fb64ffde79e37522cf555c3deb98a2831988126c/frontend/packages/client/src/hooks/useCommunityProposalsWithVotes.js#L7-L8) this endpoint returns up to 100 elements, then at the top level we can filter which ids we have and set the prop as admin or not.

germanurrus commented on frontend/packages/client/src/pages/Community.js:
```diff
@@ -194,7 +194,7 @@ export default function CommunityPage() {
         {!loading && (
           <div className="columns m-0 p-0">
-            <div className="column p-0">
+            <div className="column p-0" style={{ maxWidth: "100%" }}>
```
question here: was this breaking somehow that needed to be added?

germanurrus reviewed (changes requested):
Looks good! just left some comments

germanurrus commented:
@CJamesKeller not sure why but looks like some comments are duplicated, maybe github issues?

CJamesKeller commented on frontend/packages/client/src/pages/Community.js:
```diff
@@ -194,7 +194,7 @@ export default function CommunityPage() {
         {!loading && (
           <div className="columns m-0 p-0">
-            <div className="column p-0">
+            <div className="column p-0" style={{ maxWidth: "100%" }}>
```
This was probably initially due to the width calculation for the text wrapping. Removed.

CJamesKeller commented on frontend/packages/client/src/components/ProposalsList/ProposalCard.js:
```diff
@@ -0,0 +1,75 @@
+        <MainContent />
+      </div>
+    </Link>
+  );
```
Good call -- updated the file structure for the Proposal List & Card.

CJamesKeller commented on frontend/packages/client/src/components/CommunityProposals.js:
```diff
@@ -53,7 +53,7 @@ export default function CommunityProposals({ communityId = 1 } = {}) {
 
   return (
-    <div className={`columns ${classesContainer}`}>
+    <div className={`columns mx-0 ${classesContainer}`}>
```
This was probably initially due to the width calculation for the text wrapping. Removed.

germanurrus reviewed (changes requested):
it's looking good Chris! I made some comments, thanks for the changes

dbslone commented on frontend/packages/client/src/components/CommunityProposalList.js:
```diff
@@ -50,47 +46,24 @@ const CommunityProposalList = ({
+            <ProposalCard
+              pr={pr}
+              key={i}
+              isAdminProposal={admins.includes(pr.creatorAddr)}
```
`isAdminProposal` is not being used here so the prop can be removed

germanurrus replied:
I think `isAdminProposal' is used as a prop for `pr` here so needs to be passed done as a field for `pr` or as a prop for the component: 
used here https://github.com/DapperCollectives/CAST/pull/28/files/6613c6fe7894a2ecf5317d4f794fc422b774be6c#diff-a66b5e3ef6d31d017d9ce500657909be134fdf3b1bf05ac3547c03d487cbf005R7
passed here: 
https://github.com/DapperCollectives/CAST/pull/28/files/6613c6fe7894a2ecf5317d4f794fc422b774be6c#diff-49e211276445b84a1fd28ba2576e9eddd238fd568ebd8c0db04630400363ccb1R17
from this main component: 
https://github.com/DapperCollectives/CAST/pull/28/files/6613c6fe7894a2ecf5317d4f794fc422b774be6c#diff-30382faeb0dc980e175c3fc2fececb79b5b0df137d195f465f905d4c66daf50aR6
maybe pass this as a separate prop form the main content so it's not lost in the prop down drilling?

CJamesKeller replied:
Shouldn't have missed that! Updated.

dbslone commented on frontend/packages/client/src/components/CommunityProposalList.js:
```diff
@@ -50,47 +46,24 @@ const CommunityProposalList = ({
-                </div>
-              </div>
-            </Link>
+              isAdminProposal={admins.includes(pr.creatorAddr)}
```
same as above, remove the prop
